lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,472 @@
+	// becomes extremely important in _connectAtomicTxnsWrapper().
+	var totalInnerTxnFees uint64
+	var err error
+	for _, innerTxn := range txn.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns {
+		totalInnerTxnFees, err = SafeUint64().Add(totalInnerTxnFees, innerTxn.TxnFeeNanos)
+		if err != nil {
+			return RuleErrorAtomicTxnsWrapperHasInternalFeeOverflow
+		}
+	}
+	if txn.TxnFeeNanos != totalInnerTxnFees {
+		return RuleErrorAtomicTxnsWrapperMustHaveEqualFeeToInternalTxns
+	}
+
+	// Technically the txn.TxnNonce field could be anything but for consistent
+	// hashing we force all fields of the nonce to be zero. This also makes
+	// it more consistent with the rest of the rules regarding atomic transactions
+	// wrappers.
+	if txn.TxnNonce.ExpirationBlockHeight != 0 || txn.TxnNonce.PartialID != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroedNonce
+	}
+
+	// NOTE: We do not enforce rules on txn.ExtraData as it's both useful
+	// 		 for app developers and is being paid for via txn.TxnFeeNanos.
+
+	return nil
+}
+
+func _verifyAtomicTxnsChain(txnMeta *AtomicTxnsWrapperMetadata) error {
+	// Validate:
+	//	(1) The inner transactions are not additional redundant atomic transactions wrappers.
+	//  (2) The inner transactions are meant to be included in an atomic transaction.
+	//	(3) The start point is the first inner transaction and there's only one start point.
+	// We also collect the atomic hash of each inner transaction here for convenience.
+	var atomicHashes []*BlockHash
+	for ii, innerTxn := range txnMeta.Txns {
+		// Validate this transaction is not another redundant atomic transaction.
+		if innerTxn.TxnMeta.GetTxnType() == TxnTypeAtomicTxnsWrapper {
+			return RuleErrorAtomicTxnsHasAtomicTxnsInnerTxn
+		}
+
+		// Validate the inner transaction as meant to be included in an atomic transaction.
+		if !innerTxn.IsAtomicTxnsInnerTxn() {
+			return RuleErrorAtomicTxnsHasNonAtomicInnerTxn
+		}
+
+		// Validate the starting point of the atomic transactions chain.
+		_, keyExists := innerTxn.ExtraData[AtomicTxnsChainLength]
+		if !keyExists && ii == 0 {
+			return RuleErrorAtomicTxnsMustStartWithChainLength
+		}
+		if keyExists && ii > 0 {
+			return RuleErrorAtomicTxnsHasMoreThanOneStartPoint
+		}
+
+		// The error check in AtomicHash() is almost redundant, but we must keep it in the event
+		// that the byte buffer for the Sha256 hash fails to allocate. This should almost never
+		// occur, and there's more serious issues if it does.
+		innerTxnAtomicHash, err := innerTxn.AtomicHash()
+		if err != nil {
+			return errors.Wrap(err, "_verifyAtomicTxnsChain")
+		}
+		atomicHashes = append(atomicHashes, innerTxnAtomicHash)
+	}
+
+	// Construct special helper functions for circular doubly linked list indexing.
+	nextIndex := func(currentIndex int, chainLength int) int {
+		// Check for the special case of an atomic chain of length 1.
+		if chainLength == 1 {
+			return currentIndex
+		}
```

Comment:
Isn't this unnecessary really? If `chainLength == 1` then any (integer + 1) % 1 will be 0.